### PR TITLE
Run integration tests against a Postgres container (Testcontainers)

### PR DESCRIPTION
**Pull request by Dalet**
## Motivation
Integration tests against an in-memory db are not reliable: some SQL requests can work perfectly against the in-memory db and fail on the real database. This PR aims to make running the tests against a real database very simple (no more docker-compose).
## Changes summary
- Changed the README to mention the Docker dependency and how to run in-memory tests

### Test project:
- Made tests run against a temporary Postgres Docker container via the [Testcontainers library](https://dotnet.testcontainers.org/)
  - The container is created only once and automatically deleted at the end.
  - The database is reset by dropping the db and recreating it from an already seeded template DB.
- Changed the setup of some test classes to run only once for faster execution time.

### CI/build changes:
- Removed db-test from docker-compose because it is replaced by Testcontainers (and its associated `POSTGRES_TEST_PORT` env variable from `example.env`
- Made .env file completely optional (.csproj edits), and removed the step that creates it in the CI.
- Separated nuget restore, build, and test steps on the CI so it is easier to see where failure occurs.
- Removed in-memory tests (it's still possible to run them with this env variable: `INTEGRATION_TESTS_DB_BACKEND=InMemory`)

### New dependencies:
- [Testcontainers](https://dotnet.testcontainers.org/)
- Docker

## Performance
I've made resetting the database as fast as I could, but keep in mind resetting the db is slow, and often easily avoidable.

API requests in `[SetUp]` also take a lot of time if there are a lot of tests in the class, so they should be kept to a minimum as well.

### CODE REVIEW

File: LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		}
+
+		PostgresContainer = new ContainerBuilder<PostgreSqlTestcontainer>()
+			.WithDatabase(new PostgreSqlTestcontainerConfiguration
+			{
+				Database = "db",
+				Username = "dbuser",
+				Password = "weft5gst768sr",
+			})
+			.WithTmpfsMount("/var/lib/postgresql/data") // db files in-memory
```
Comment (RageCage64) [inline LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs:36]: Is this mount on the container or the system? Cause if it's on the system we'll need to make this work on Windows too.

**Review by RageCage64 — COMMENTED** [review]
If we always have to do the same setup in every test fixture, can we make a base class? Maybe something like `APIResourceTestFixture`?

File: LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs (hunk 1 above)
Comment (Dalet) [reply LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs:36]: this is on the container

Comment (Dalet) [reply LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs:36]: Actually I'm not sure, but I tested it on Windows.

File: README.md [hunk 2]
```diff
@@ -145,6 +145,8 @@ dotnet run --project LeaderboardBackend --urls https://localhost:7128 // dotnet
 The value provided to `--urls` has to match what's listed under `applicationUrl` in `LeaderboardBackend/Properties/launchSettings.json`. As of this writing, it's `https://localhost:7128`.
 
 #### Test the App
+Docker is required to run integration tests against a real database.  
+However, if you would like to run the tests with an in-memory database, you can set the following environment variable: `INTEGRATION_TESTS_DB_BACKEND=InMemory`.
```
Comment (zysim) [inline README.md:149]: Since you've mentioned that we can have situations where tests can work if we use an in-memory DB, but fail in a real one, do we want to mention that here? Maybe encourage contributors to always test with a real DB instance before requesting review?

File: LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+	[OneTimeSetUp]
+	public static async Task OneTimeSetup()
+	{
+		if (TestConfig.DatabaseBackend != DatabaseBackend.TestContainer)
+		{
+			return;
+		}
+
+		PostgresContainer = new ContainerBuilder<PostgreSqlTestcontainer>()
```
Comment (zysim) [inline LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs:29]: Github Actions are warning that this constructor's obsolete. Mind fixing this accordingly?

Comment (zysim) [inline LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs:29]: Github Actions are warning that this constructor's obsolete. Mind fixing this accordingly?

File: LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+		NpgsqlConnection.ClearAllPools(); // can't drop a DB if connections remain open
+		using NpgsqlDataSource conn = CreateConnectionToTemplate();
+		conn.CreateCommand(@$"
+			DROP DATABASE IF EXISTS {TemplateDatabase};
+			CREATE DATABASE {TemplateDatabase}
+				WITH TEMPLATE {PostgresContainer!.Database}
+				OWNER '{PostgresContainer!.Username}';
+			")
+			.ExecuteNonQuery();
```
Comment (zysim) [inline LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs:64]: Sorry but I'm having a hard time following this. So `CreateConnectionToTemplate()` returns a DataSource that's connected to `PostgresContainer`, but with its database set to "template1". But `conn.CreateCommand()` would create a database named "db_template", wouldn't it? If so, do we use "template1" anywhere?

**Review by zysim — CHANGES_REQUESTED** [review]

**Comment by Dalet** [comment]
> If we always have to do the same setup in every test fixture, can we make a base class? Maybe something like `APIResourceTestFixture`?
Absolutely, I think it's a good idea but outside the scope of the PR.

File: LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs (hunk 3 above)
Comment (Dalet) [reply LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs:29]: https://github.com/testcontainers/testcontainers-dotnet/issues/750#issuecomment-1412257694
It wasn't possible to fix it until the next version of Testcontainers would release... which [it finally did 4 hours ago](https://github.com/testcontainers/testcontainers-dotnet/releases/tag/3.0.0)!
Will fix in the coming days.

Comment (zysim) [reply LeaderboardBackend.Test/Fixtures/PostgresDatabaseFixture.cs:29]: Ah okay. Sweet

**Review by zysim — APPROVED** [review]

**Comment by RageCage64** [comment]
> > If we always have to do the same setup in every test fixture, can we make a base class? Maybe something like `APIResourceTestFixture`?
> 
> Absolutely, I think it's a good idea but outside the scope of the PR.

I personally think it's in scope since this is the PR that adds the setup that matches a clear potential abstraction. But I don't feel strongly enough to block anything.

**Review by zysim — APPROVED** [review]

**Comment by zysim** [comment]
> > > If we always have to do the same setup in every test fixture, can we make a base class? Maybe something like `APIResourceTestFixture`?
> > 
> > 
> > Absolutely, I think it's a good idea but outside the scope of the PR.
> 
> I personally think it's in scope since this is the PR that adds the setup that matches a clear potential abstraction. But I don't feel strongly enough to block anything.

Let's just make that change in its own PR. Don't foresee any clashes with other PRs and stuff.
